Feature element workflow

JoshuaCarter opened the pull request:
Add 2 new actions:
- `SelectElementsAction` for selecting elements via the right sidebar.
- `PublishWithSelectedElements` for publishing **only** the selected elements on the page (rather than all).

Additionally addresses:
- `update()` was being called on actions when the page was saved (no actions were using). The result was that `DataObject::update()` was called and it tried to write the Page data into the action, resulting in a bunch of junk entries in the DB table, etc.
- Adds composer requirements.
- Fixes some styling issues with the right sidebar. In particular that it was overlapping the edit form and hiding element controls, etc.
- Brings some config out of the project as defaults in the module.

Notes:

- I haven't looked into 'per elements change diffing' yet. That whole thing looks like it could use some improvement.

Commit:
mod(composer): Expose whole client folder

Commit:
add(admin): Style and config

Commit:
add(admin): Align checkbox for element selection

Commit:
fix(WorkflowFieldCapture): Replace update() called on actions

- update() replaced with onSaveWorkflowPage()
- pre/post SaveWorkflowPage methods ensure parent methods are called
  (think onBeforeWrite).
- No actions were using update() but DataObject has an update()
  method which was attempting to write the supplied Page fields
  into the action's DB table, resulting in many junk rows with the
  Page's ID.

Commit:
mod(config): Move some workflow config into module

Also apply WorkflowActionInstanceExtension

Commit:
dept(sidebar): Add missing requirements

- advancedworkflow
- multivaluefield
- rightsidebar

Commit:
fix(WorkflowedElement): Move warning banner above Title field

Commit:
fix(Rightsidebar): Sidebar no longer overlaps with Root edit form

Commit:
add(action): New SelectElements action/instance and styling

This will be the 'choose which elements are part of this workflow and
will be published' action.

Commit:
add(sidebar): Readonly selected elements list

Shows selected elements list in the sidebar when there have
previously been elements selected.

Commit:
fix(sidebar): Top padding now static

Was previously dependant on the h2 title being 1 line.

Commit:
add(action): PublishWithSelectedElementsAction

As an alternative to recursively publishing all elements

Commit:
chore(comments): Add comments to recent work

No functional changes

nyeholt commented on client/styles/admin.css:
```diff
@@ -0,0 +1,74 @@
+/* SelectElement */
+.wfa-right-elements-list {
```
Does this _expect_ the right side bar config to be in place?

JoshuaCarter replied:
No, that's a custom class I add here:

https://github.com/nyeholt/silverstripe-workflow-actions/pull/1/files#diff-e4e683a478ef409169197f3975a28218R78

nyeholt commented on src/Extension/WorkflowFieldCapture.php:
```diff
@@ -12,14 +12,18 @@ class WorkflowFieldCapture extends DataExtension
                 $action = $active->CurrentAction();
                 if ($action) {
-                    $action->update($this->owner->toMap());
+                    $postVars = Controller::curr()->getRequest()->postVars();
```
Does the default behaviour here still do what the previous behaviour did, whereby _all_ fields that come through the submit get written to the object?

JoshuaCarter replied:
It's more that this wasn't doing what was intended to begin with I think. The previous code is writing a **Page**'s fields into the **Action** object/table. Thus:

- The update() uses the Page's ID to find an action, which either:
  - Fails and a new Action with the Page's ID is created.
  - Or it finds an action made as above.
  - Or it finds a valid action who's ID happens to be the same as the Page (and which will now be overwritten, invalidating it).
- The update() then writes any Page fields that overlap with the Action (ID, Created, etc). If you look at the WorkflowActionInstance table, you'll see lots of entries missing their ClassName, those are the failed attempts to write a Page into that table.
- IIRC Comment had, by that point, been wrongly assigned to the Page object, and so it was present in the `toMap()` call.
- Comment is now written into the action via `WorkflowActionInstanceExtension::onSaveWorkflowPage()`
- It's worth noting that no action was overriding the update() method.

It's very possible I've missed something in my implementation of all this, but either way the previous approach needs fixing.

nyeholt replied:
Yeah, that does look busted; I believe the original functionality was done that way because when a user hits 'save' or the _next_ workflow step button, anything entered in the right hand form needs to get saved into the workflow action instance (ie the Comment field, or any other field that action expects). Likely what needs to happen is that the `WorkflowActionInstanceExtension` needs to look at all the DB fields of the action instance _except a reserved list_ and set those values if they're present on the request. IIRC there's no way of accessing the form for the request by the time the actions are being triggered, so the only way to access the data (without grabbing the controller directly) is off the page being modified, which _does_ have the workflow action fields set against it directly on save. 

In its current state, this will break existing workflows that have additional properties being collected in the right column.

JoshuaCarter replied:
I think we should discuss this IRL for clarity and speed.

JoshuaCarter replied:
Solution discussed IRL now implemented. Key point is that we're now using `invokeWithExtensions()`.

JoshuaCarter commented on src/Extension/WorkflowFieldCapture.php:
```diff
@@ -12,14 +12,18 @@ class WorkflowFieldCapture extends DataExtension
 {
     public function onBeforeWrite()
     {
+        // TODO: make this check for instance
```
I need to fix this!

JoshuaCarter replied:
done

Commit:
fix(WorkflowFieldCapture): Allow saving without curr controller

And minor reformat of checks

Commit:
fix(WorkflowFieldCapture): Use invokeWithExtensions() over direct call

- Removes need for calling parent method and makes it easier to extend
  actions with data.

nyeholt reviewed (approved):
Looks good!
